[Merged by Bors] -  chore(*): add a `div`/`sub` field to (`add_`)`group`(`_with_zero`)

Vierkantor opened the pull request:
This PR is intended to fix the following kind of diamonds:
Let `foo X` be a type with a `∀ X, has_div (foo X)` instance but no `∀ X, has_inv (foo X)`, e.g. when `foo X` is a `euclidean_domain`. Suppose we also have an instance `∀ X [cromulent X], group_with_zero (foo X)`. Then the `(/)` coming from `group_with_zero_has_div` cannot be defeq to the `(/)` coming from `foo.has_div`.

As a consequence of making the `has_div` instances defeq, we can no longer assume that `(div_eq_mul_inv a b : a / b = a * b⁻¹) = rfl` for all groups. The previous preparation PR #5302 should have changed all places in mathlib that assumed defeqness, to rewrite explicitly instead.

---

- [x] depends on: #5302

Vierkantor committed:
chore(*): add a `div`/`sub` field to (`add_`)`group`(`_with_zero`)

This PR is intended to fix the following kind of diamonds:
Let `foo X` be a type with a `∀ X, has_div (foo X)` instance but no `∀ X, has_inv (foo X)`,
   e.g. when `foo X` is a `euclidean_domain`.
Suppose we also have an instance `∀ X [cromulent X], group_with_zero (foo X)`.
Then the `(/)` coming from `group_with_zero_has_div` cannot be defeq to the `(/)` coming from `foo.has_div`.

As a consequence of making the `has_div` instances defeq, we can no longer
assume that `(div_eq_mul_inv a b : a / b = a * b⁻¹) = rfl` for all groups.
The previous preparation commit should have changed all places in mathlib
that assumed defeqness, to rewrite explicitly instead.

Vierkantor committed:
Fix overlong lines

Vierkantor committed:
More `sub_eq_add_neg` rewriting

Vierkantor committed:
Fill in missing field so the elaborator doesn't complain about metavariables

Vierkantor committed:
Fix `analysis/hofer.lean`

Vierkantor committed:
Fix remaining issues

Vierkantor committed:
Lint fix (missing docstring)

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -290,16 +290,73 @@ class cancel_comm_monoid (M : Type u) extends left_cancel_comm_monoid M, right_c
-(add_left_neg : ∀ a : α, -a + a = 0)
+/-- `div_sub_tac` proves `∀ a b, a / b = a * b⁻¹` or `∀ a b, a - b = a + -b`
+if it holds by definition. -/
+meta def div_sub_tac : tactic unit := `[intros; refl]
```
I think this prevents it occupying a name in the global namespace?
```suggestion
private meta def div_sub_tac : tactic unit := `[intros; refl]
```

Vierkantor replied:
I'm inclined to keep it non-private since I expect it is more likely to be reused rather than cause name collisions. (The corresponding `order_laws_tac` for `preorder` is also non-private.)

digama0 replied:
frankly these definitions don't need to be namespaced at all, it can just be called `try_refl_tac` instead of having multiple identical tactics. (Also the position after `.` does not do name resolution, so you have to spell out all the namespaces, and if you used `private` it wouldn't be accessible at all.)

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -290,16 +290,73 @@ class cancel_comm_monoid (M : Type u) extends left_cancel_comm_monoid M, right_c
+class group (G : Type u) extends div_inv_monoid G :=
+(mul_left_inv : ∀ a : G, a⁻¹ * a = 1)
+
+attribute [to_additive group.to_sub_neg_monoid] group.to_div_inv_monoid
```
Should this be:
```suggestion
attribute [to_additive add_group.to_sub_neg_monoid] group.to_div_inv_monoid
```
Can `to_additive` not work it out automatically?

Vierkantor replied:
Good point, I think we can just delete that line entirely since the definition of `add_group` would generate it anyway.

Vierkantor committed:
This line is probably not needed?

sgouezel commented on src/algebra/group_with_zero/basic.lean:
```diff
@@ -770,7 +765,7 @@ lemma div_eq_inv_mul : a / b = b⁻¹ * a :=
 
 lemma mul_div_right_comm (a b c : G₀) : (a * b) / c = (a / c) * b :=
-by rw [div_eq_mul_inv, mul_assoc, mul_comm b, ← mul_assoc, div_eq_mul_inv]
+by { rw [div_eq_mul_inv, mul_assoc, mul_comm b, ← mul_assoc, div_eq_mul_inv] }
```
Why would you change this?

Vierkantor replied:
I probably added an extra tactic call and forgot to delete the brackets when I deleted the tactic.
```suggestion
by rw [div_eq_mul_inv, mul_assoc, mul_comm b, ← mul_assoc, div_eq_mul_inv]
```

sgouezel reviewed (commented):
It would be worth adding somehwere a comment explaining why you choose to have fields `sub` and `div` instead of using `x + - y` or `x * y^{-1}`, by mentioning the example that is a motivation for the PR.

Vierkantor committed:
Update src/algebra/group_with_zero/basic.lean

Vierkantor committed:
Explain the motivation behind `div_inv_monoid`

Vierkantor committed:
Can we delete the `int_sub_hack`?

Vierkantor committed:
Fix `congruence.lean`

Vierkantor committed:
Rename `div_sub_tac` -> `try_refl_tac`

Vierkantor committed:
Fix overlong line

(I could have sworn the line was just short enough; I cannot count apparently)

Vierkantor committed:
Fix missing `div_eq_mul_inv`

The elaborator managed to apply it by itself in a really messed-up way,
so it looked like `norm_num` was unfolding stuff badly when really it was the
elaborator a few lines previously
